### PR TITLE
[BEAM-4461] CoGroup transforms for schemas.

### PR DESCRIPTION
**Pull request by reuvenlax**
This provides the base for a schema-based join transform. Higher-level join transforms will be layered on top of this.

### CODE REVIEW

File: sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/transforms/CoGroupTest.java [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
+
+/** Tests for {@link CoGroup}. */
+public class CoGroupTest {
+  @Rule public final transient TestPipeline pipeline = TestPipeline.create();
+  @Rule public transient ExpectedException thrown = ExpectedException.none();
+
+  private static final Schema JOIN_SCHEMA_1 =
```
Comment (akedin) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/transforms/CoGroupTest.java]: I don't think we should implicitly conflate CoGbk with Join. I think people think of join result as a collection of rows, i.e. cogbk_output+merge_into_single_row = join

**Comment by reuvenlax** [comment]
Run Java PreCommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java [hunk 2]
```diff
@@ -209,6 +209,25 @@ public boolean equals(Object o) {
...
+  /** Returns true if two schemas are equal ignoring field names and descriptions. */
+  public boolean typesEqual(Schema other) {
+    if (uuid != null && other.uuid != null && Objects.equals(uuid, other.uuid)) {
+      return true;
+    }
+    if (!Objects.equals(fieldIndices.values(), other.fieldIndices.values())) {
+      return false;
+    }
+    if (getFieldCount() != other.getFieldCount()) {
+      return false;
```
Comment (akedin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java]: would it make sense to check the field count first before checking the `fieldIndices` equality?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java [hunk 3]
```diff
@@ -474,6 +493,29 @@ public boolean equals(Object o) {
           && Arrays.equals(getMetadata(), other.getMetadata());
     }
 
+    /** Returns true if two FieldTypes are equal, ignoring name and description. */
```
Comment (akedin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java]: It doesn't seem to ignore the type name, copy-paste error?

**Comment by akedin** [comment]
run java precommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java (hunk 2 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java (hunk 3 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java]: Confusing terminology - "TypeName" is the type, not the field name.

**Comment by akedin** [comment]
run java precommit

**Pull request merged**